packages/actor_pinning/actor_pinning.pony

```diff
@@ -0,0 +1,85 @@
+"""
+# Actor Pinning Package
+
+The Actor Pinning package allows Pony programmers to pin actors to a dedicated 
+scheduler thread. This can be required/used for interfacing with C libraries
+that rely on thread local storage. A common example of this is graphics/windowing
+libraries.
+
+The way it works is that an actor can request that it be pinned and then check
+to confirm that the pinning was successfully applied after which all subsequent
+behaviors on that actor will run on the same scheduler thread until the actor is
+destroyed or the actor requests to be unpinned.
+
+## Example program
+
+```pony
+// Here we have the Main actor that upon construction requests a PinUnpinActorAuth
+// token from AmbientAuth and then requests that it be pinned. It then recursively
+// calls the `check_pinned` behavior until the runtime reports that it has
+// successfully been pinned after which it starts `do_stuff` to do whatever
+// work it needs to do that requires it to be pinned. Once it has completed all
+// of its work, it calls `done` to request that the runtime `unpin` it.
```

Comment:
talk with Red and Adrian during office hours, I think this needs to be elaborated on as it requires some intimate knowledge of scheduler for this to make sense.

Reply:
The gist that they didnt get is...

that if you start a scheduler run on a non-pinned actor thread and pin yourself, you won't be running on the pinned actor thread until a new run is started so it is important to check to see if you are on the "pinned actor thread".

I'm not sure of the best way to explain this without getting into scheduler/runtime internals.

Reply:
i don't follow.. the API exposed doesn't require any knowledge of how the scheduler works.. it does imply that `pinning` an actor is not necessarily an immediate action and might take some time to go into effect but doesn't preclude it from happening immediately (in theory, the compiler/language could be modified pass enough info to `pony_create` such that the actor could be marked as `pinned` upon creation for example but that might also suggest a rethinking of the API)..

maybe you or Red or Adrian can elaborate on why you feel it requires intimate knowledge of the scheduler?

Reply:
also, it is possible to make the `pin` go into effect immediately with additional scheduler changes if that is desired (have `ActorPinning.pin` signal to the scheduler that this actor should stop processing additional behaviors after the existing one ends but i haven't thought through the performance considerations on non-`pinned` actor workloads of adding this additional logic)..

Reply:
I think going into effect immediately has a larger performance impact.

I think "request_pin" is a good idea.

Reply:
doing it immediately would add an additional branch after every behavior run at:

https://github.com/ponylang/ponyc/blob/f0d7afdd0b86190f917c3c6a8883082ae2ce862f/src/libponyrt/actor/actor.c#L622

to check to see if the `actor` needs to yield it's normal batch of behaviors to run early or not..

this is likely to be negligible impact from a performance perspective for non-`pinned` actors due to the branch predictor magic in modern cpus (i think)..

let me know if that changes your perspective on whether to rename the API function to `request_pin` or to instead make changes to have the `pin` go into effect immediately (and also to remove `is_successfully_pinned` function from the API)..

Reply:
we discussed some more and they think that "request" would have made it more clear to them that it isn't immediate.

I think what we want to convey is "this won't happen immediately and any time you run, it isn't guaranteed you are on the 'pin thread'".

Reply:
i think if we convey the above then we can safely safe "how this happens and when is implementation dependent". we need to convey that "in the future" and "you need to check".

Reply:
yep.. agreed..

Reply:
updates the package function names and docs and release notes..